packages/sdk-components-react/src/vimeo.tsx

```diff
@@ -350,6 +224,115 @@ export type VimeoOptions = Omit<
   showPortrait?: VimeoPlayerOptions["portrait"];
 };
 
+const EmptyState = () => {
+  return (
+    <div
+      style={{
+        display: "flex",
+        width: "100%",
+        height: "100%",
+        alignItems: "center",
+        justifyContent: "center",
+        fontSize: "1.2em",
+      }}
+    >
+      {
+        'Open the "Settings" panel and paste a video URL, e.g. https://vimeo.com/831343124.'
+      }
+    </div>
+  );
+};
+
+type PlayerStatus = "initial" | "loading" | "ready";
+
+type PlayerProps = Pick<
+  VimeoOptions,
+  "loading" | "autoplay" | "showPreview"
+> & {
+  videoUrl: string;
+  status: PlayerStatus;
+  renderer: ContextType<typeof ReactSdkContext>["renderer"];
+  previewImageUrl?: URL;
+  onStatusChange: (status: PlayerStatus) => void;
+  onPreviewImageUrlChange: (url?: URL) => void;
+};
+
+const Player = ({
+  status,
+  loading,
+  videoUrl,
+  previewImageUrl,
+  autoplay,
+  renderer,
+  showPreview,
+  onStatusChange,
+  onPreviewImageUrlChange,
+}: PlayerProps) => {
+  const [opacity, setOpacity] = useState(0);
+
+  useEffect(() => {
+    if (autoplay && renderer !== "canvas" && status === "initial") {
+      onStatusChange("loading");
+    }
+  }, [autoplay, status, renderer, onStatusChange]);
+
+  useEffect(() => {
+    if (videoUrl === undefined) {
+      return;
+    }
+
+    if (showPreview === false) {
+      onPreviewImageUrlChange(undefined);
+      return;
+    }
+
+    if (previewImageUrl === undefined) {
+      loadPreviewImageUrl(videoUrl)
+        .then(onPreviewImageUrlChange)
```

Comment:
Usually this is bug in react as effect can be reexecuted and outdated `onPreviewImageUrlChange` will be called
Any async with callbacks in React are bugs

Reply:
yeah, potentially, is there an easy way to do fetch from a component?